### PR TITLE
Drop support for series in bundles

### PR DESCRIPTION
**Pull request by jack-w-shaw**
Drop support for series in bundles

Support for series in bundles is being dropped in 4.0 (major version
bump). Error out if bundles include series

This means we no longer need to parse series into base in the rest of
our bundle deploy code

We do not want to rely the juju/charm BundleData struct to parse series.
We want to keep this data structure clean, so will be removing from
BundleData very soon Series

As such, we need to parse the bundle ourselves. This requires using a
new miner version of "juju/charm" which allows us to access the raw
unparsed bundle

We can also drop compatibility parsing in bundle handlers

Also, rename `checkExplicitSeries` to `checkExplicitBase` and stop checking
if `series` is provided here.

Also, swap our series for base in test bundles

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [x] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

### Attempt to deploy some bundles 

(Note: charmed-kubernetes rev 1243 contains series on the top level bundle, no where else)
```
$ juju deploy charmed-kubernetes --revision 1243
Located bundle "charmed-kubernetes" in charm-hub, revision 1243
ERROR cannot deploy bundle: the provided bundle has the following errors:
base bundle contains invalid key series:
- document 0; bundle contains top level series
```

```
$ cat ./bundle.yaml
default-base: ubuntu@22.04/stable
applications:
  ubuntu:
    charm: ubuntu
    num_units: 1
    to:
    - "0"
machines:
  "0":

$ juju deploy ./bundle.yaml
Located charm "ubuntu" in charm-hub, channel latest/stable
Executing changes:
- upload charm ubuntu from charm-hub for base ubuntu@22.04/stable with architecture=amd64
- deploy application ubuntu from charm-hub on ubuntu@22.04/stable
- add new machine 0
- add unit ubuntu/0 to new machine 0
Deploy of bundle completed.
```
```
# We should also be able to deploy by specifying a directory with a bundle
$ juju deploy .
Located charm "ubuntu" in charm-hub, channel latest/stable
Executing changes:
- upload charm ubuntu from charm-hub for base ubuntu@22.04/stable with architecture=amd64
- deploy application ubuntu from charm-hub on ubuntu@22.04/stable
- add new machine 0
- add unit ubuntu/0 to new machine 0
Deploy of bundle completed.
```

```
$ cat ./bundle.yaml
series: jammy
applications:
  ubuntu:
    charm: ubuntu
    num_units: 1
    to:
    - "0"
machines:
  "0":

$ juju deploy ./bundle.yaml
ERROR cannot deploy bundle: the provided bundle has the following errors:
base bundle contains invalid key series:
- document 0; bundle contains top level series
```

```
$ cat bundle.yaml 
series: jammy
applications:
  ubuntu:
    charm: ubuntu
    series: jammy
    num_units: 1
    to:
    - "0"
machines:
  "0":
    series: jammy

$ juju deploy ./bundle.yaml
ERROR cannot deploy bundle: the provided bundle has the following errors:
base bundle contains invalid key series:
- document 0; bundle contains top level series
- docuemnt 0; bundle application "ubuntu" contains series
- document 0; bundle machine "0" contains series
```

### Attempt to deploy some bundles with overlays

```
$ cat bundle.yaml
series: jammy
applications:
  ubuntu:
    charm: ubuntu
    num_units: 1
    to:
    - "0"
machines:
  "0":
---
applications:
  ubuntu:
    series: jammy
machines:
  "0":
    series: jammy

$ juju deploy ./bundle.yaml
ERROR cannot deploy bundle: the provided bundle has the following errors:
base bundle contains invalid key series:
- document 0; bundle contains top level series
- docuemnt 1; bundle application "ubuntu" contains series
- document 1; bundle machine "0" contains series
```

```
$ cat bundle.yaml
applications:
  ubuntu:
    charm: ubuntu
    num_units: 1
    to:
    - "0"
machines:
  "0":

$ cat overlay1.yaml
applications:
  ubuntu:
    series: jammy
machines:
  "0":
    series: jammy

$ juju deploy ./bundle.yaml --overlay ./overlay1.yaml
ERROR cannot deploy bundle: the provided bundle has the following errors:
overlay index 0 contains invalid key series:
- document 0; bundle application "ubuntu" contains series
- document 0; bundle machine "0" contains series
```

```
$ cat bundle.yaml
applications:
  ubuntu:
    charm: ubuntu
    num_units: 1
    to:
    - "0"
machines:
  "0":

$ cat overlay1.yaml
applications:
  ubuntu:
    series: jammy
machines:
  "0":
    series: jammy

$ cat overlay2.yaml
applications:
  ubuntu:
    series: jammy
machines:
  "0":
    series: jammy

$ juju deploy ./bundle.yaml --overlay ./overlay1.yaml --overlay ./overlay2.yaml
ERROR cannot deploy bundle: the provided bundle has the following errors:
overlay index 0 contains invalid key series:
- docuemnt 0; bundle application "ubuntu" contains series
- document 0; bundle machine "0" contains series
overlay index 1 contains invalid key series:
- docuemnt 0; bundle application "ubuntu" contains series
- document 0; bundle machine "0" contains series
```

## Links

**Jira card:** [JUJU-5996](https://warthogs.atlassian.net/browse/JUJU-5996)

[JUJU-5996]: https://warthogs.atlassian.net/browse/JUJU-5996?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: cmd/juju/application/bundle/bundle.go [hunk 1]
```diff
@@ -330,14 +330,7 @@ func verifyBundle(data *charm.BundleData, bundleDir string, verifyConstraints fu
...
 
-	var errs []string
-	// This method cannot be included within data.Verify because
-	// to verify corresponding series and base match we need to be
-	// able to compare them. The charm package, however, treats bases
-	// and series generically and is unable to do this.
-	if err := verifyMixedSeriesBasesMatch(data); err != nil {
-		errs = append(errs, err.Error())
-	}
+	errs := assertNoSeries(data)
```
Comment (SimonRichardson) [inline cmd/juju/application/bundle/bundle.go:333]:
```suggestion
	errs := verifyNoSeries(data)
```

or `ensureNoSeries`.

**Review by SimonRichardson — COMMENTED** [review]
TBH: I think we just remove Series from the bundle code and nothing happens it's just ignored.

**Comment by jack-w-shaw** [comment]
> TBH: I think we just remove Series from the bundle code and nothing happens it's just ignored.

I don't have any problems with that, but i would like to hear more voices at the go around. 

A have added another commit which makes this change

**Comment by jack-w-shaw** [comment]
> TBH: I think we just remove Series from the bundle code and nothing happens it's just ignored.

We have discussed this together further and landed on erroring out if series is provided, but parsing for this in `juju/juju`

File: cmd/juju/application/bundle/bundle.go [hunk 2]
```diff
@@ -350,59 +345,74 @@ func verifyBundle(data *charm.BundleData, bundleDir string, verifyConstraints fu
...
+	Machines map[string]struct {
+		Series string `yaml:"series"`
+	} `yaml:"machines"`
+}
+
+func verifyNoSeries(dslist []charm.BundleDataSource) []string {
+	errs := []string{}
+
+	// The first dslist is always the bundle itself
+	bundleErrs := verifyBundleNoSeries(dslist[0].RawBundle())
```
Comment (SimonRichardson) [inline cmd/juju/application/bundle/bundle.go:372]: I know that there is a check above where this is called to check that dslist is empty, but I don't like this coding style. Instead, can this return an empty string if the list is empty.

File: cmd/juju/application/bundle/bundle.go [hunk 3]
```diff
@@ -350,59 +345,74 @@ func verifyBundle(data *charm.BundleData, bundleDir string, verifyConstraints fu
...
+	for docIdx := 0; ; docIdx++ {
+		var data *bundleSeriesData
+
+		err := dec.Decode(&data)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			// The bundle should already have been parsed if we're
+			// calling this check, so we should not see this
+			return []string{fmt.Sprintf("programming error: %w", err)}
```
Comment (SimonRichardson) [inline cmd/juju/application/bundle/bundle.go:400]: Remove `programming error`. I'd just return the error, things can be refactored and seeing `programming error` is not ideal.

**Review by SimonRichardson — COMMENTED** [review]
This is good to go once we change the method name from `RawBytes()`

File: cmd/juju/application/bundle/bundle.go (hunk 2 above)
Comment (jack-w-shaw) [reply cmd/juju/application/bundle/bundle.go:372]: You mean like a catch guard? Sure thing

File: cmd/juju/application/bundle/bundle.go (hunk 3 above)
Comment (jack-w-shaw) [reply cmd/juju/application/bundle/bundle.go:400]: Done

**Review by SimonRichardson — APPROVED** [review]

**Pull request merged**